tools/pmunts_muntsos/config-pin.c

```diff
@@ -49,6 +50,80 @@ static void fixup_pin_name(char *s)
   }
 }
 
+// Only stores the first line of cmd's output
+// Returns -1 if fails
+static int RunShellCmd(char *cmd, char *buff, const int BUFF_SIZE) {
+  FILE *fp;
+  int status;
+
+  fp = popen(cmd, "r");
+  if(fp == NULL)
+    return -1;
+  
+  fgets(buff, BUFF_SIZE, fp);
+  strtok(buff, "\n"); // Remove newline
+
+  status = pclose(fp);
+
+  return status;
+}
+
+// Returns -1 if fails
+static int GetGpio(char *pin) {
+  char cmd[255];
+  char pin_name[32];
+  char chip[2];
+  char gpio_chip[3];
+  int gpio, status;
+
+  fixup_pin_name(pin);
+
+  // Ensure pin name is valid (prevent OS injections)
+  // (must have format P[0-9]_[0-9][0-9])
+
+  if(strlen(pin) != 5 || pin[0] != 'P' || !isdigit(pin[1]) || pin[2] != '_' || !isdigit(pin[3]) || !isdigit(pin[4]))
+  {
+    fprintf(stderr, "WARNING: invalid pin name, %s", pin,
+    strerror(errno));
+    return -1;
+  }
+
+  // Get pin name
+
+  strcpy(cmd, "gpioinfo | grep ");
+  strcat(cmd, pin);
+  strcat(cmd, " -m 1 | awk '{print substr($3,2,length($3) - 2)}'");
```

Comment:
One thing to consider here is doing the post-processing here in C (instead of with awk). Your checks above protect against injection attacks pretty well, but the less stuff you do in subprocesses, the better.

Reply:
I have removed the calls to awk as recommended, see commits 8dba275 and 695b9b6 . The main concern was the first call to awk (because this is where the user's input is used), but I have removed the other calls as well. Doing this allows the code to only invoke _popen_ twice instead of three times.